pallets/messages/src/lib.rs

```diff
@@ -315,28 +345,25 @@ impl<T: Config> Pallet<T> {
 		msa_id: Option<MessageSourceId>,
 		payload: BoundedVec<u8, T::MaxMessagePayloadSizeBytes>,
 		schema_id: SchemaId,
-	) -> Result<Message<T::MaxMessagePayloadSizeBytes>, DispatchError> {
-		<BlockMessages<T>>::try_mutate(
-			|existing_messages| -> Result<Message<T::MaxMessagePayloadSizeBytes>, DispatchError> {
-				let current_size: u16 = existing_messages
-					.len()
-					.try_into()
-					.map_err(|_| Error::<T>::TypeConversionOverflow)?;
-
-				let msg = Message {
-					payload, // size is checked on top of extrinsic
-					provider_msa_id,
-					msa_id,
-					index: current_size,
-				};
-
-				existing_messages
-					.try_push((msg.clone(), schema_id))
-					.map_err(|_| Error::<T>::TooManyMessagesInBlock)?;
-
-				Ok(msg)
-			},
-		)
+		current_block: T::BlockNumber,
+	) -> DispatchResult {
+		<Messages<T>>::try_mutate(current_block, schema_id, |existing_messages| -> DispatchResult {
+			let index = MessageIndex::<T>::get();
+			let msg = Message {
+				payload, // size is checked on top of extrinsic
+				provider_msa_id,
+				msa_id,
+				index,
+			};
+
+			existing_messages
+				.try_push(msg)
+				.map_err(|_| Error::<T>::TooManyMessagesInBlock)?;
+
+			Self::track_schema_id(schema_id)?;
```

Comment:
you can avoid a `try_mutate` on ```BlockSchema``` if ```existing_messages count``` is ==1 is when you need to add schema id to  ```BlockSchema``` otherwise we know can safely assume it exists and avoid calls to storage on each message add?

Reply:
I see you are adding a counter too, probably there is a need

Reply:
To Avoid changing event contract I'm counting how many times it's used. Nevermind since it seems like `whitelist_storage` is not working as expected and we have to figure out what is going on there